README.md
# pyChakraCore

pyChakraCore is a Python encapsulation of ChakraCore which is a C project. ChakraCore is a JavaScript engine that Microsoft Edge browser ever used.

[ [README-EN](pyChakraCore/README_EN.md)]

[ [中文说明](pyChakraCore/README_CN.md)]

[ [CHANGELOG](pyChakraCore/CHANGELOG.md)]


pyChakraCore/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 0.2.0 (2021-03-22)

### ⚠ BREAKING CHANGES

-   new API,more API.See in [README-EN#API](https://github.com/wistn/pyChakraCore/blob/main/pyChakraCore/README_EN.md#API)

### Features

-   Improve codes ([9528fca](https://github.com/wistn/pyChakraCore/commit/9528fca253256cd960236a2fe8032c28cb6727d8))
-   new API,more API.See in [README-EN#API](https://github.com/wistn/pyChakraCore/blob/main/pyChakraCore/README_EN.md#API) ([74ae933](https://github.com/wistn/pyChakraCore/commit/74ae933e46d6affdcd7c179bb07ff213dbb9f31b))

# 0.1.0 (2020-10-20)

first version, public


test.py
# -*- coding: UTF-8 -*-
"""
Author:wistn
since:2020-10-20
LastEditTime:,:2020-10-20
LastEditors:,:Do not edit
Description:
"""
import sys
import os
# if __package__ == "" or __package__ == None:
#     path = os.path.dirname(os.path.dirname(__file__))
#     sys.path.insert(0, path)
from pyChakraCore import pyChakraCore


if sys.argv.__len__() > 1 and os.path.isfile(sys.argv[1]) and ".js" in sys.argv[1]:
    with open(sys.argv[1], "r", encoding="utf-8") as fs:
        jscode = fs.read()
else:
    jscode = "console.log(typeof (1 + 1) == 'number')"
vm = pyChakraCore()
vm.run(jscode)



pyChakraCore/test.py
# -*- coding: UTF-8 -*-
"""
Author:wistn
since:2021-03-14
LastEditors:Do not edit
LastEditTime:2021-03-16
Description:
"""
import unittest
import os
import sys

if __package__ == "" or __package__ == None:
    path = os.path.join(os.path.dirname(os.path.realpath(__file__)), "..")
    sys.path.insert(0, path)
from pyChakraCore import pyChakraCore, failCheck

# print(sys.modules["pyChakraCore"])
testObj = pyChakraCore()
# the class below cann't be defined in any function if uses unittest.main method.
class TestCase(unittest.TestCase):
    def setUp(self):
        self.vm = testObj

    def tearDown(self):
        pass  # Method called immediately after the test method has been called and the result recorded, to cleanup resources used during the test like writes 'self.widget.dispose()'. This method will only be called if the internal setUp() succeeds, regardless of the outcome of the test method.

    def test_callFunction(self):
        ret = self.vm.callFunction("String", "1+1")
        # print("test_callFunction", "2")
        self.assertEqual(ret, "2")

    def test_callFunctionToJson(self):
        ret = self.vm.callFunctionToJson("new Array", "'str1'")
        # print("test_callFunctionToJson", repr('["str1"]'))
        self.assertEqual(ret, '["str1"]')

    def test_getJValue(self):
        ret = self.vm.getJValue("undefined")
        # print("test_getJValue", type(ret))
        self.assertEqual(ret.value, self.vm.jUndefined.value)

    def test_getValueType(self):
        ret = self.vm.getValueType(self.vm.jNull)
        # print("test_getValueType", ret)
        self.assertEqual(ret, "JsNull")

    def test_jGet_jSetProperty(self):
        failCheck(
            self.vm.jSetProperty(
                self.vm.getJValue("console"),
                "alsoLog",
                self.vm.jGetProperty(self.vm.getJValue("console"), "log"),
            )
        )
        ret = self.vm.run("console.alsoLog")
        # print("test_jGet_jSetProperty", ret)
        self.assertEqual(ret, "function () { [native code] }")

    def test_jValueToNativeInt(self):
        ret = self.vm.jValueToNativeInt(self.vm.getJValue("1"))
        # print("test_jValueToNativeInt", type(ret))
        self.assertEqual(ret, 1)

    def test_jValueToNativeStr(self):
        ret = self.vm.jValueToNativeStr(self.vm.getJValue("1"))
        # print("test_jValueToNativeStr", type(ret))
        self.assertEqual(ret, "1")

    def test_registerMethod(self):
        self.vm.registerMethod(self.vm.c_EchoCallback, "console.info")
        ret = self.vm.run("console.info('test_registerMethod','hi')")
        self.assertEqual(ret, "undefined")

    def test_run(self):
        ret = self.vm.run("console.log('test_run: typeof 1 is', typeof 1)")
        self.assertEqual(ret, "undefined")


def main():
    fullname = os.path.realpath(__file__)
    unittest.main(
        module=os.path.basename(os.path.dirname(fullname))
        + "."
        + os.path.basename(fullname).replace(".py", ""),
        exit=False,
        verbosity=2,
    )
    """the following commented-out code equals unittest.main(module="PackageName.ModuleName", exit=True, verbosity=2)"""
    # suite = unittest.TestSuite()
    # arr = [
    #     TestCase("test_run"),
    #     TestCase("test_getJValue"),
    #     TestCase("test_callFunction"),
    #     TestCase("test_callFunctionToJson"),
    #     TestCase("test_registerMethod"),
    #     TestCase("test_getValueType"),
    #     TestCase("test_jValueToNativeInt"),
    #     TestCase("test_jValueToNativeStr"),
    #     TestCase("test_jGet_jSetProperty"),
    # ]
    # suite.addTests(arr)
    # runner = unittest.TextTestRunner(verbosity=2)
    # runner.run(suite)


if __name__ == "__main__":
    main()
    pass


pyChakraCore/ChakraCore.py
# -*- coding: UTF-8 -*-
"""
Author:wistn
since:2020-12-20
LastEditors:Do not edit
LastEditTime:2021-03-15
Description:
"""

import ctypes
import os
import sys
import platform
import types

if sys.platform == "darwin":
    pathChakraCore = os.path.join(
        os.path.split(os.path.realpath(__file__))[0], "osx", "libChakraCore.dylib"
    )
elif sys.platform.startswith("linux"):
    pathChakraCore = os.path.join(
        os.path.split(os.path.realpath(__file__))[0], "linux", "libChakraCore.so"
    )
elif sys.platform == "win32":
    if platform.architecture()[0].startswith("64"):
        pathChakraCore = os.path.join(
            os.path.split(os.path.realpath(__file__))[0], "windowsx64", "ChakraCore.dll"
        )
    else:
        pathChakraCore = os.path.join(
            os.path.split(os.path.realpath(__file__))[0], "windowsx86", "ChakraCore.dll"
        )

if sys.platform != "win32":
    ChakraCore = ctypes.CDLL(pathChakraCore)
else:
    ChakraCore = ctypes.WinDLL(pathChakraCore)
if sys.platform != "win32":
    # You are expected to call DllMain manually on non-Windows
    ChakraCore.DllMain(0, 1, 0)  # Attach process
    ChakraCore.DllMain(0, 2, 0)  # Attach main thread


def JsGetPropertyIdFromName(self, name, ptr_propertyId):
    return ChakraCore.JsCreatePropertyId(
        name.encode("UTF-8"), len(name.encode("UTF-8")), ptr_propertyId
    )


ChakraCore.JsGetPropertyIdFromName = types.MethodType(
    JsGetPropertyIdFromName, ChakraCore
)


def JS_INVALID_REFERENCE():
    # 要函数创建，不能直接赋值变量 JS_INVALID_REFERENCE = ctypes.c_void_p()，否则其他变量引用后被ctypes.byref操作后会影响 JS_INVALID_REFERENCE 的值。
    return ctypes.c_void_p()


class AutoRestoreContext:
    def __init__(self, newContext):
        self.oldContext = ctypes.c_void_p()
        self.contextChanged = False
        ChakraCore.JsGetCurrentContext(ctypes.byref(self.oldContext))
        if self.oldContext.value != newContext.value:
            ChakraCore.JsSetCurrentContext(newContext)
            self.contextChanged = True
        else:
            self.contextChanged = False

    def __del__(self):
        if (
            self.contextChanged
            and self.oldContext.value != JS_INVALID_REFERENCE().value
        ):
            ChakraCore.JsSetCurrentContext(self.oldContext)


.github/workflows/changelog_new_section.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
"""
Author:wistn
since:2021-03-11
LastEditors:Do not edit
LastEditTime:2021-03-16
Description:
"""
import os
import re

project_root = os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", "..")
with open(
    os.path.join(project_root, "pyChakraCore", "CHANGELOG.md"), "r", encoding="utf-8"
) as fs:
    text = fs.read().replace("\r\n", "\n")
regexp = r"(^|\n)(#+\s+\[?v?\d(\.\d){0,2}\]?[\s\S]*?)\n#+ \[?v?\d(\.\d){0,2}\]?"
if re.search(regexp, text, re.I):
    new_section = re.search(regexp, text, re.I)[2]
    """ like:
        ## [2.0.0]
        'second_section'
        # v1
    """
else:
    new_section = re.search(r"(^|\n)(#+\s+\[?v?\d(\.\d){0,2}\]?[\s\S]*)$", text, re.I)[
        2
    ]
    """ like:
        # Changelog
        ### 0.0.1
        'first_section'
    """
version = re.search(r"#+\s+\[?v?(\d(\.\d){0,2})\]?", new_section, re.I)[1]
with open(os.path.join(project_root, "new_section"), "w", encoding="utf-8") as fs:
    fs.write(new_section)
with open(os.path.join(project_root, "version"), "w", encoding="utf-8") as fs:
    fs.write(version)


pyChakraCore/__init__.py
# -*- coding: UTF-8 -*-
"""
Author:wistn
since:2020-08-03
LastEditors:Do not edit
LastEditTime:2021-03-16
Description:
"""
from .ChakraCore import *
from .ChakraCommon import *
import ctypes
import json

__version__ = "0.2.0"
pretty = r"""
    var backslashN = '\n';
    function pretty(obj, space) {
        var indent = '',
            subIndents = '';
        if (space == null) space = 4;
        if (typeof space == 'number') {
            for (var i = 0; i < space; i++) {
                indent += ' ';
            }
        } else if (typeof space == 'string') {
            indent = space;
        }
        function str(obj) {
            var jsType = Object.prototype.toString.call(obj);
            if (
                jsType.match(/object (String|Date|Function|JSON|Math|RegExp)/)
            ) {
                return JSON.stringify(String(obj));
            } else if (jsType.match(/object (Number|Boolean|Null)/)) {
                return JSON.stringify(obj);
            } else if (jsType.match(/object Undefined/)) {
                return JSON.stringify('undefined');
            } else {
                if (jsType.match(/object (Array|Arguments|Map|Set)/)) {
                    if (jsType.match(/object (Map|Set)/)) {
                        /* es6新增的方法和参数类型 */
                        obj = Array.from(obj);
                    }
                    var partial = [];
                    subIndents = subIndents + indent;
                    var len = obj.length;
                    for (var i = 0; i < len; i++) {
                        partial.push(str(obj[i]));
                    }
                    var result =
                        len == 0
                            ? '[]'
                            : indent.length
                            ? '[' +
                              backslashN +
                              subIndents +
                              partial.join(',' + backslashN + subIndents) +
                              backslashN +
                              subIndents.slice(indent.length) +
                              ']'
                            : '[' + partial.join(',') + ']';
                    subIndents = subIndents.slice(indent.length);
                    return result;
                } else if (
                    jsType.match(
                        /object (Object|Error|global|Window|HTMLDocument)/i
                    ) ||
                    obj instanceof Error
                ) {
                    var partial = [];
                    subIndents = subIndents + indent;
                    var ownProps = Object.getOwnPropertyNames(obj);
                    /* Object.keys 为自身非继承属性(不用for in因为for遍历继承的祖先属性)，Object.getOwnPropertyNames 在前者基础上包括不可枚举属性  */
                    var len = ownProps.length;
                    for (var i = 0; i < len; i++) {
                        partial.push(
                            str(ownProps[i]) +
                                (indent.length ? ': ' : ':') +
                                str(obj[ownProps[i]])
                        );
                    }
                    var result =
                        len == 0
                            ? '{}'
                            : indent.length
                            ? '{' +
                              backslashN +
                              subIndents +
                              partial.join(',' + backslashN + subIndents) +
                              backslashN +
                              subIndents.slice(indent.length) +
                              '}'
                            : '{' + partial.join(',') + '}';
                    subIndents = subIndents.slice(indent.length);
                    return result;
                } else {
                    return JSON.stringify(String(obj));
                }
            }
        }
        function decycle(obj) {
            /* the function can solve circular structure like JSON.decycle, the return value can be decoded by JSON.retrocycle(JSON.parse()) */
            var arrParents = [];
            return (function derez(obj, path) {
                var jsType = Object.prototype.toString.call(obj);
                if (
                    jsType.match(
                        /object (String|Date|Function|JSON|Math|RegExp|Number|Boolean|Null|Undefined)/
                    )
                ) {
                    return obj;
                } else {
                    if (jsType.match(/object (Array|Arguments|Map|Set)/)) {
                        var len = arrParents.length;
                        for (var i = 0; i < len; i++) {
                            /* arr like [obj, '$'] */
                            var arr = arrParents[i];
                            if (obj === arr[0]) {
                                return { $ref: arr[1] };
                            }
                        }
                        arrParents.push([obj, path]);
                        var newObj = [];
                        if (jsType.match(/object (Map|Set)/)) {
                            /* es6新增的方法和参数类型 */
                            obj = Array.from(obj);
                        }
                        var length = obj.length;
                        for (var i = 0; i < length; i++) {
                            newObj[i] = derez(obj[i], path + '[' + i + ']');
                        }
                        return newObj;
                    } else {
                        var len = arrParents.length;
                        for (var i = 0; i < len; i++) {
                            /* arr like [obj, '$'] */
                            var arr = arrParents[i];
                            if (obj === arr[0]) {
                                return { $ref: arr[1] };
                            }
                        }
                        arrParents.push([obj, path]);
                        var newObj = {};
                        var ownProps = Object.getOwnPropertyNames(obj);
                        var length = ownProps.length;
                        for (var i = 0; i < length; i++) {
                            newObj[ownProps[i]] = derez(
                                obj[ownProps[i]],
                                path + '[' + JSON.stringify(ownProps[i]) + ']'
                            );
                        }
                        return newObj;
                    }
                }
            })(obj, '$');
        }
        return str(decycle(obj));
    }
"""


class pyChakraCore:
    def __del__(self):
        ChakraCore.JsSetCurrentContext(self.context)
        ChakraCore.JsDisposeRuntime(self.runtime)

    def __init__(self):
        self.runtime = ctypes.c_void_p()  # Create Javascript Runtime.
        ChakraCore.JsCreateRuntime(
            0x00000000, ctypes.c_void_p(), ctypes.byref(self.runtime)
        )
        self.context = ctypes.c_void_p()
        ChakraCore.JsCreateContext(
            self.runtime, ctypes.byref(self.context)
        )  # Create an execution context.一个Runtime可包含多个Context
        self.j_sourceContext = 0
        ChakraCore.JsSetCurrentContext(self.context)
        # Now set the current execution context.
        self.jGlobal = ctypes.c_void_p()
        ChakraCore.JsGetGlobalObject(ctypes.byref(self.jGlobal))
        # 避免展开WebAssembly对象时失败，所以直接删除对象。
        ChakraCore.JsDeleteProperty(
            self.jGlobal,
            pyChakraCore.getPropertyIdFromName("WebAssembly"),
            False,
            ctypes.byref(ctypes.c_void_p()),
        )
        self.jUndefined = ctypes.c_void_p()
        ChakraCore.JsGetUndefinedValue(ctypes.byref(self.jUndefined))
        self.jNull = ctypes.c_void_p()
        ChakraCore.JsGetNullValue(ctypes.byref(self.jNull))
        self.run(pretty)
        self.c_EchoCallback = cfunctype_JsNativeFunction(self.EchoCallback)
        self.registerMethod(self.c_EchoCallback, "print")
        # self.registerMethod(self.c_EchoCallback, "WScript.Echo")
        self.registerMethod(self.c_EchoCallback, "console.log")

    def run(self, code):
        j_result = self.getJValue(code)
        self.j_sourceContext += 1
        return pyChakraCore.jValueToNativeStr(j_result)

    def getJValue(self, code):
        ChakraCore.JsSetCurrentContext(self.context)
        j_sourceContext = self.j_sourceContext
        sourceUrl = "sourceUrl:" + code.strip()[0:15]
        j_sourceUrl = ctypes.c_void_p()  # create JsValueRef from sourceUrl
        ChakraCore.JsCreateString(
            sourceUrl.encode("UTF-8"),
            len(sourceUrl.encode("UTF-8")),
            ctypes.byref(j_sourceUrl),
        )
        cBuffer_scriptSource = ctypes.create_string_buffer(code.encode("UTF-16"))
        j_scriptSource = ctypes.c_void_p()  # Create ArrayBuffer from script
        ChakraCore.JsCreateExternalArrayBuffer(
            cBuffer_scriptSource,
            len(cBuffer_scriptSource),
            ctypes.c_void_p(),
            ctypes.c_void_p(),
            ctypes.byref(j_scriptSource),
        )
        j_result = ctypes.c_void_p()
        failCheck(
            ChakraCore.JsRun(
                j_scriptSource,
                j_sourceContext,
                j_sourceUrl,
                0x2,
                ctypes.byref(j_result),
            )
        )
        return j_result

    def _jValueToJson(self, jValue):
        j_result = self._callFunctionWithJArgs(
            "JSON.stringify", (ctypes.c_void_p * 2)(self.jUndefined, jValue),
        )
        strJSON = pyChakraCore.jValueToNativeStr(j_result)
        return strJSON

    def _jValueToPretty(self, jValue):
        j_result = self._callFunctionWithJArgs(
            "pretty", (ctypes.c_void_p * 2)(self.jUndefined, jValue),
        )
        return pyChakraCore.jValueToNativeStr(j_result)

    def callFunction(self, funName, *code_arguments):
        argsArray_code = "[" + ",".join(code_arguments) + "]"
        return self.run(funName + "(..." + argsArray_code + ")")
        # Do not use func.apply because not know the value of this provided for the call to func

    def _callFunctionWithJArgs(self, funName, j_arguments):
        j_result = ctypes.c_void_p()
        jFunction = self.getJValue(funName)
        # 或者global.funName调用
        # j_result = ctypes.c_void_p()
        # components = funName.split(".")
        # jObj = self.jGlobal
        # for i in range(len(components)):
        #     jValue = pyChakraCore.jGetProperty(jObj, components[i])
        #     jObj = jValue
        # jFunction = jValue
        failCheck(
            ChakraCore.JsCallFunction(
                jFunction,
                ctypes.byref(j_arguments),
                len(j_arguments),
                ctypes.byref(j_result),
            )
        )
        return j_result

    def _callConstructorWithJArgs(self, funName, j_arguments):
        j_result = ctypes.c_void_p()
        funName = funName.strip()
        if "new " in funName:
            funName = funName[4:].strip()
        jFunction = self.getJValue(funName)
        failCheck(
            ChakraCore.JsConstructObject(
                jFunction,
                ctypes.byref(j_arguments),
                len(j_arguments),
                ctypes.byref(j_result),
            )
        )
        return j_result

    def callFunctionToJson(self, funName, *code_arguments):
        argsArray_code = "[" + ",".join(code_arguments) + "]"
        strJSON = self.run("JSON.stringify(" + funName + "(..." + argsArray_code + "))")
        return strJSON

    def registerMethod(
        self, nativeFunction, funName, callbackState=ctypes.py_object(None)
    ):
        ChakraCore.JsSetCurrentContext(self.context)
        jFunction = ctypes.c_void_p()
        ChakraCore.JsCreateFunction(
            nativeFunction, callbackState, ctypes.byref(jFunction),
        )  # 匿名函数
        components = funName.split(".")
        jObj = self.jGlobal
        for i in range(len(components)):
            if i == len(components) - 1:
                pyChakraCore.jSetProperty(jObj, components[i], jFunction, True)
                break
            jValue = pyChakraCore.jGetProperty(jObj, components[i])
            if pyChakraCore.getValueType(jValue) == "JsUndefined":
                jValue = ctypes.c_void_p()
                ChakraCore.JsCreateObject(ctypes.byref(jValue))
                pyChakraCore.jSetProperty(jObj, components[i], jValue, True)
            jObj = jValue

    def EchoCallback(
        self, ptr_callee, isConstructCall, ptrj_arguments, argumentCount, callbackState,
    ):
        # chakraCore虚拟机里，注册方法第一个参数都是undefined。当其他参数出现not defined 时，不会进入本原生函数，而是JsRun返回空字符串''回调用处。
        for i in range(1, argumentCount):
            jString = ctypes.c_void_p(ptrj_arguments[i])
            string = pyChakraCore.jValueToNativeStr(jString)
            if i > 1:
                print(" ", end="")
            print(string, end="")
        print("\n", end="")
        # return self.jUndefined.value


pyChakraCore.getValueType = staticmethod(getValueType)
pyChakraCore.jValueToNativeInt = staticmethod(jValueToNativeInt)
pyChakraCore.jValueToNativeStr = staticmethod(jValueToNativeStr)
pyChakraCore.getPropertyIdFromName = staticmethod(getPropertyIdFromName)
pyChakraCore.jGetProperty = staticmethod(jGetProperty)
pyChakraCore.jSetProperty = staticmethod(jSetProperty)

